File: Embassat/View/Artist Detail/CADEMArtistDetailViewController.m
//
//  CADEMArtistDetailViewController.m
//  Embassat
//
//  Created by Joan Romano on 26/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMArtistDetailViewController.h"

#import "CADEMArtistDetailViewModel.h"
#import <SDWebImage/UIImageView+WebCache.h>

@interface CADEMArtistDetailViewController ()

@property (nonatomic, weak) IBOutlet UIScrollView *scrollView;;
@property (nonatomic, weak) IBOutlet UIImageView *coverImage;
@property (nonatomic, weak) IBOutlet UIView *infoView;
@property (nonatomic, weak) IBOutlet UILabel *descriptionLabel, *dayLabel, *timeLabel, *stageLabel;
@property (nonatomic, weak) IBOutlet UIButton *shareButton, *addButton;

@end

@implementation CADEMArtistDetailViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    @weakify(self)
    
    self.scrollView.scrollIndicatorInsets = (UIEdgeInsets){
        .top = CGRectGetHeight(self.coverImage.bounds) + CGRectGetHeight(self.infoView.bounds),
        .bottom = CGRectGetHeight(self.addButton.bounds)
    };
    
    self.descriptionLabel.font = [UIFont em_romanFontOfSize:16.0f];
    self.stageLabel.font = self.dayLabel.font = self.timeLabel.font = self.shareButton.titleLabel.font = self.addButton.titleLabel.font = [UIFont em_boldFontOfSize:16.0f];
    
    self.title = [self.viewModel.artistName uppercaseString];
    self.view.backgroundColor = [UIColor whiteColor];

    [self.coverImage sd_setImageWithURL:self.viewModel.artistImageURL placeholderImage:[UIImage imageNamed:@"placeholder.jpg"]];
    self.stageLabel.text = self.viewModel.artistStage;
    self.dayLabel.text = self.viewModel.artistDay;
    self.descriptionLabel.text = self.viewModel.artistDescription;
    self.timeLabel.text = [NSString stringWithFormat:@"%@.%@", self.viewModel.artistStartHour, self.viewModel.artistStartMinute];
    
    [[self.addButton rac_signalForControlEvents:UIControlEventTouchUpInside] subscribeNext:^(id x) {
        @strongify(self)
        [self.viewModel addEventOnCalendar];
    }];
}

@end

File: Embassat/View/Artist Detail/CADEMArtistDetailViewController.swift
//
//  CADEMArtistDetailViewController.swift
//  Embassa't
//
//  Created by Joan Romano on 17/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import UIKit
import ReactiveCocoa
import WebImage

public class CADEMArtistDetailViewController: CADEMRootViewController {
    
    @IBOutlet weak var scrollView: UIScrollView?
    @IBOutlet weak var coverImage: UIImageView?
    @IBOutlet weak var infoView: UIView?
    @IBOutlet weak var bottomView: UIView?
    @IBOutlet weak var artistNameLabel: UILabel?
    @IBOutlet weak var descriptionLabel: UILabel?
    @IBOutlet weak var dayLabel: UILabel?
    @IBOutlet weak var timeLabel: UILabel?
    @IBOutlet weak var stageLabel: UILabel?
    
    public var updateSignal: RACSubject = RACSubject()
    public var viewModel: CADEMArtistDetailViewModel? {
        didSet {
            self.updateSubviewDetails()
        }
    }

    override public func viewDidLoad() {
        super.viewDidLoad()
        
        title = "Artista"
        artistNameLabel?.font = UIFont.detailFont(ofSize: 20.0)
        descriptionLabel?.font = UIFont.detailFont(ofSize: 15.0)
        stageLabel?.font = UIFont.detailFont(ofSize: 15.0)
        dayLabel?.font = UIFont.detailFont(ofSize: 15.0)
        timeLabel?.font = UIFont.detailFont(ofSize: 15.0)
        
        let shareItem = UIBarButtonItem(image: UIImage(named: "share.png"), style: UIBarButtonItemStyle.Plain, target: self, action: #selector(CADEMArtistDetailViewController.sharePressed))
        let favItem = UIBarButtonItem(image: UIImage(named: "fav.png"), style: UIBarButtonItemStyle.Plain, target: self, action: #selector(CADEMArtistDetailViewController.favoritePressed))
        
        self.navigationItem.rightBarButtonItems = [favItem, shareItem]
    }
    
    @IBAction func nextPressed(sender: UIButton) {
        viewModel?.currentIndex+=1
        self.updateSubviewDetails()
    }
    
    @IBAction func previousPressed(sender: UIButton) {
        viewModel?.currentIndex-=1
        self.updateSubviewDetails()
    }
    
    func updateSubviewDetails() {
        artistNameLabel?.text = viewModel?.artistName
        descriptionLabel?.text = viewModel?.artistDescription
        coverImage?.sd_setImageWithURL(viewModel?.artistImageURL, placeholderImage: UIImage(named: "loading.jpg"))
        stageLabel?.text = viewModel?.artistStage
        dayLabel?.text = viewModel?.artistDay
        timeLabel?.text = viewModel?.artistStartTimeString
        
        let favItem = self.navigationItem.rightBarButtonItems?.first
        favItem?.tintColor = viewModel?.artistIsFavorite == true ? UIColor.emBackgroundColor() : nil
    }
    
    func sharePressed() {
        viewModel?.shareAction(forViewController: self)
    }
    
    func favoritePressed() {
        viewModel?.toggleFavorite(){ [unowned self] () -> () in

            let favItem = self.navigationItem.rightBarButtonItems?.first
            favItem?.tintColor = favItem?.tintColor == nil ? UIColor.emBackgroundColor() : nil
            
            self.updateSignal.sendNext(true)
        }
    }
}
